twc/src/main/java/org/openmbee/sdvc/twc/security/TwcAuthenticationFilter.java

```diff
@@ -0,0 +1,63 @@
+package org.openmbee.sdvc.twc.security;
+
+import java.io.IOException;
+import javax.servlet.FilterChain;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import org.openmbee.sdvc.core.utils.RestUtils;
+import org.openmbee.sdvc.permissions.users.UserDetailsImpl;
+import org.openmbee.sdvc.permissions.users.UserDetailsServiceImpl;
+import org.openmbee.sdvc.twc.config.TwcConfig;
+import org.openmbee.sdvc.twc.constants.TwcConstants;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.security.authentication.AuthenticationManager;
+import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.security.web.authentication.WebAuthenticationDetailsSource;
+import org.springframework.security.web.authentication.www.BasicAuthenticationFilter;
+
+
+public class TwcAuthenticationFilter extends BasicAuthenticationFilter {
+
+
+    @Autowired
+    private UserDetailsServiceImpl userDetailsService;
+
+	@Autowired
+	private TwcConfig twcConfig;
+
+
+	public TwcAuthenticationFilter(AuthenticationManager authenticationManager) {
+		super(authenticationManager);
+	}
+
+	@Override
+	protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain)
+			throws ServletException, IOException {
+
+
+		try {
+		    String associatedTWC = request.getHeader(TwcConstants.ASSOCIATED_TWC_HEADER);
+		    String token = request.getHeader(RestUtils.AUTHORIZATION);
+		    TwcAuthenticationProvider twcAuthProvider = twcConfig.getAuthNProvider(associatedTWC);
+		    if(twcAuthProvider != null) {
+                String authenticatedUser = twcAuthProvider.getAuthentication(token);
+                if(authenticatedUser != null) {
+                    UserDetailsImpl userDetails = this.userDetailsService.loadUserByUsername(authenticatedUser);
```

Comment:
Our use cases require 

- authentication from multiple sources (From within Cameo, from a SSO provider, etc)
- projects from multiple sources (Cameo and other)

These all need to work together, so our goal is to decouple the Authentication and Authorization.  
There could be multiple sources of Authentication (TWC would just be one of them).   
The source of Authorization for each project/branch would be defined by project/branch attributes.  
Some projects could use the internal permissions scheme, and others could delegate to TWC for Authorization.

Reply:
Right, for the first one (multiple authentication sources), that's already provided by spring security's filter mechanism - it'll go through all filters until one of them sets the authentication object - I'm just confused by the TwcAuthenticationFilter as to why after getting the user from the TWC Auth token, it's loading the user from the local database? This would only succeed if the local database gets synced with TWC users somehow, but I don't see that happening anywhere.

For the permissions - it'd be possible to do something similar I think, but it's not there yet - basically go through all possible permissionService implementations until it provides an answer - or have an implementation that does what you describe

Reply:
I think the user will still need to exist locally.  If the user doesn't exist locally, but is authenticated by TWC, we should probably add them automatically.  I'll add that in.

I'm still thinking about how to integrate the permissions, but I like the idea of multiple permission service implementations.

Reply:
Well, it's not strictly necessary for the user to exist locally for auth - that may be a convenience if we want to add user profiles/settings down the line - but that would be up to you - be aware though if you load the user details from the db it'll have no authorities unless you also sync those (if there are any that should be attached), otherwise you can also just return them from the TwcUserDetailsImpl directly

Reply:
There is no need to sync the authorities with TWC since we're going to delegate those back to TWC anyway.  I agree that it can still work without adding the user to the db, but I think we still want to add them for the sake of consistency.  We will still want to pick up MMSADMIN rights for users that authenticate via TWC tokens